docs/utilities/parser.md

```diff
@@ -0,0 +1,216 @@
+| **snsSqsEnvelope**            | 1. Parses data using `SqsSchema`. <br/> 2. Parses SNS records in `body` key using `SnsNotificationSchema`. <br/> 3. Parses data in `Message` key using your schema and return them in a list.                 |
+| **sqsEnvelope**               | 1. Parses data using `SqsSchema`. <br/> 2. Parses records in `body` key using your schema and return them in a list.                                                                                          |
+| **vpcLatticeEnvelope**        | 1. Parses data using `VpcLatticeSchema`. <br/> 2. Parses `value` key using your schema and returns it.                                                                                                        |
+| **vpcLatticeV2Envelope**      | 1. Parses data using `VpcLatticeSchema`. <br/> 2. Parses `value` key using your schema and returns it.                                                                                                        |
+
+
+## Manual parsing
+
+You can use built-in envelopes and schemas to parse the incoming events manually: 
+
+=== "Manual parsing"
+    ```typescript hl_lines="25 28"
+    --8<-- "docs/snippets/parser/manual.ts"
+    ```
+    
+    1. Use `EventBridgeSchema` to parse the event, the `details` fields will be parsed as a generic record.
+    2. Use `eventBridgeEnvelope` with a combination of `orderSchema` to get `Order` object from the `details` field.
+
+
+## Custom validation
+
+Because Parser uses zod, you can use all the features of zod to validate your data.
+For example, you can use `refine` to validate a field or a combination of fields:
+
+=== "Custom validation"
+    ```typescript hl_lines="13 18"
+    --8<-- "docs/snippets/parser/refine.ts"
+    ```
+
+Zod provides a lot of other features and customization, see [zod documentation](https://zod.dev) for more details.
+
+
+## Types
+
+### Schema and Type inference
+Zod provides a way to extract the type of schema, so you can use it in your handler:
+
+=== "Types"
+    ```typescript hl_lines="19 22 27"
+    --8<-- "docs/snippets/parser/types.ts"
+    ```
+    
+    1. Use `z.infer` to extract the type of the schema
+    2. `event` is typed as `Order` object
+    3. we can infer deeply nested types, here `Order` and `OrderItem` are inferred
+
+### Compatibility with @types/aws-lambda
+
+The package `@types/aws-lambda` contains type definitions for AWS service event invocations. 
+Powertools parser utility also bring AWS Lambda event types based on the built-in schema definitions. 
+Both libraries try to provide types for the same event structure, but they might not be compatible with each other yet. 
+We are working on a sustainable solution to make them compatible and avoid any breaking changes in the future. 
```

Comment:
Reading this I am a bit confused (as a customer), mainly because I don't understand whether this is an issue with Powertools or `@types/aws-lambda`.

As a maintainer, while I know we have an issue open for this, I think we should agree & articulate our stance on this topic a bit better to avoid future confusion amongst us and towards customers.

The main point here, besides calling out potential discrepacies between us & the other library, is in my opinion to establish our strategy towards event types as a whole. This will then influence which actions we'll take now and in the future, as well as when discrepacies are detected.

Currently the `@types/aws-lambda` types are community maintained and manually authored. AWS events can change and the types can contain issues.

While I agree with the sentiment of maintaining these types and our schemas (and in the future JSON schemas as well) up to date and aligned, I don't think this ultimately guarantees that the types/schemas are actually correct.

As a think big idea, we should at least consider the option of instead investing efforts in creating automations to periodically test against real AWS events to get the actual shapes of the events sent to Lambda. Based of those, we could then derive the schemas as well as the TypeScript types.

Once we detect drift (aka changes in the events), we could then propagate them to the appropriate destinations be it within the scope of Powertools or `@types/aws-lambda`.

I am not expecting the whole industry to move over our own types (i.e. `@aws-lambda-powertools/events` or `@aws-lambda-powertools/types`) but I would still aim at becoming the source of truth for these events/schemas/types.

The DefinitelyTyped repo (aka the one that hosts `@types/*` packages) already has precedents of depending on external type definition, one of them was added in the past 6 months to the `@types/node` types to support the `fetch` module. This module was originally implemented under `undici` and later incorporated into Node.js. The source still resides in the `undici` repo, so to add support for these types they started publishing type definitions for under `undicy-types` that are now a direct depenencey pf the `@types/node` package (see [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5684af052501705829b574de2394a6c570aaed56/types/node/package.json#L19)).

While it'll be a lot of work, I believe this would bring a lot of value to customers and eventually would help with establishing a single and authoritative source of truth for AWS events that today does not exist. Doing this would benefit not only Powertools TS or the Node.js community but also Powertools Python and the Python community as a whole via Pydantic. Additionally, it would open the door to future use cases like testing factories (#445), integrations with SAM or CDK, and similar.

---

Now, back to this PR, the TL;DR; here is that at this stage and in this section I would simply call out that there could be discrepancies between `@types/aws-lambda` and our schemas. At the moment we don't have any automatic way of detecting these differences and because of this we: 1/ recommend customers to use our types whenever possible, and 2/ consider opening an issue to us whenever they detect issues.

But overall I wouldn't say that we aim at being compatible with these types because they themselves might be incorrect, and while we will definitely contribute back when finding issues, ultimately we don't have control over them.

Reply:
Ok, refreshed this part. We recommend to use out types, but if there any errors, here is the link to raise an issue. I have removed that compatibility part to not overpromise. 

I agree that we can maintain types in PT in the long run.